### DISCUSSION TITLE
Make Generic Type "where" conditions choosable

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#6807] [Language Ideas]
Until now, we can make a Generic type limited by referring "where", something like this:

```csharp
public class SomeClass<T> where T: class, SomeInterface,……
{
}
```
Now this means "T" MUST inherit from a class, with "SomeInterface".

What about make it more complicated to meet more situations, something like this following:

```csharp
public class SomeClassWithComplicatedGenericType<T> where T: (InterfaceA or InterfaceB) and InterfaceC
{
}
```
Here "and","or","not" can be used in single, or mixed with each other (something like "Pattern"), to make the functions better：

The key words can include "not"/"and"/or", the default prior should be the same as what we see in the logic symbols such as "&&","||" and '!", including "()".

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/discussions/399
https://github.com/dotnet/csharplang/discussions/4628

**Comment by HaloFour** [comment]
What does this enable you to do?  If `T` is either `InterfaceA` or `InterfaceB` it would be impossible to actually invoke any members from either of those two interfaces so there doesn't seem to be much point to enable such a constraint.

**Reply by ghost** [reply]
IMO：

At least, **"not"** is a possible key word here because sometimes we should exclude something for the generic type "T".

For **"or"**, this senario is hard to think. But sometimes we need to limit the range the generic type "T" (Only allowing "InterfaceA" or "InterfaceB"), and inside the code we can do checks to meet satisfications.

**Reply by CyrusNajmabadi** [reply]
> At least, "not" is a possible key word here because sometimes we should exclude something for the generic type "T".

Can you give an example?

**Reply by ghost** [reply]
A typical example is:

```csharp
public class SomeClass<T> where T: (not SomeInterface) //This has limited the type of "T" to exclude "T" cannot be some types
{
}
```

```csharp
public class SomeClass<T> where T: (not SomeInterface1)  and (not SomeInterface2)....
{
}
```

Although sometimes the combinations cannot be right (e.g: you CANNOT COMBINE more than one class together with "and"). But for "not" and "and" and "or", a syntax-checking can be easily found to avoid misusing.

**Reply by ghost** [reply]
We don't need to run the runtime to check by writing a lot of "if"'s……This just the aspect of language checking in syntax.

**Reply by TahirAhmadov** [reply]
@MaledongGit I think @CyrusNajmabadi wanted to see a real-life scenario where this would be useful, to get some context. (I do, too.)

**Reply by ghost** [reply]
Thank you all, sorry for my busy work and I tried my best to reply to you, the below example is just my designed thoughts about the syntax. You can have a look at my given example in practice :)

**Comment by TahirAhmadov** [comment]
I think for this to work, the often requested `if(T is IInterface) { ... }` feature needs to be available, too.
It's not trivial, though. If `where T: IInterfaceA or IInterfaceB`, and `if(T is IInterfaceA) { ... } else { ... }`, then questions arise, such as, do we "know" that `T is IInterfaceB` inside the `else` block?
Also, what about other constraints, such `class` and `struct`, `new()`, and base concrete class. Are those excluded from this feature altogether?

**Reply by ghost** [reply]
I get what you want, but this will run through runtime.
What I want……Maybe language checking better for the combinations of "and“,"or","not"?

**Comment by ghost** [comment]
Hello all：

Sorry for my busy work and rough thoughts, and I didn't make myself so clear for you（@HaloFour, @CyrusNajmabadi and @TahirAhmadov ） to understand:

Now I took some of my time to reply to you in details:

Obviously speaking, if I want to define a class with a method that needs some operators such as "+". I want to make the current types support "+" as a generic type, take this as an example:

```csharp
public static T GenericAdd<T>(T t1,T t2) where T : IAdditionOperators<T,T,T>
        {
            return t1 + t2;
        }
```

The example below has forced that those with  IAdditionOperators<T,T,T> (of course it's right with the latest Numeric interface sum supporting), But what about "string" or some customized class types supporting "+" ? (They may not implement  IAdditionOperators<T,T,T>, but still with "+" supporting...On the other hand, "string" is a sealed class....So how?)

If you say something like this, it's totally wrong because string doesn't implement the interface:

```csharp
 int number = GenericType.GenericAdd<int>(1,2); //Right, any numeric type has implemented it
 GenericType.GenericAdd<string>("1", "2"); //Wrong! string doesn't!
```

Until now, what we can do is just to decide the type of "T", something like this:
```csharp
 public static T GenericAdd<T>(T t1, T t2)
        {

            switch ((t1, t2))
            {
                case (string t11, string t22):
                    {
                        object result = t11 + t22;
                        return (T)result;
                    }
                case (int t11, int t22):
                    {
                        object result = t11 + t22;
                        return (T)result;
                    }
                // Do for other types such as float,double,decimal....
                // You cannot use IAddOperation<T,T,T>, because you never
                // know whether "T" implements IAddOperation<T,T,T>.
            }
            throw new Exception("Not implemented of add symbol yet");
        }
```

What's more: this will make you have to use a boxing/unboxing, becasuse "T" is never known what it really is.

A better solution for intellisense for VS C#'s language is to make it sensible by merging the "where" conditions together:

```csharp
  public static T GenericAdd<T>(T t1, T t2) where T:(IAdditionOperators<T,T,T> or string)
        {
            return t1 + t2;
        }
```
here, any type that supports "+" or part of the function will be checked by intellisense of VS to avoid runtime errors. Not only for the generic function, but the class should also support this function as well for some specific situations.

**Reply by ufcpp** [reply]
https://github.com/dotnet/csharplang/discussions/6308 discuss on static abstract calls

**Reply by ghost** [reply]
Thank you @ufcpp .

I'm a little diff from this idea (though most meanings are the same).  #6308 is focusing on runtime check, however mine is focusing on syntax checking (to avoid unboxing/boxing and reduce the codes).

But all in all, this enhancement is very useful, and this is just want I want :)

**Reply by HaloFour** [reply]
Your suggestion regarding addition and concatenation would not work.  The constraint `IAdditionOperators<T,T,T>` informs both the compiler and the runtime about what specific methods exist on the type that enables the use of addition.  The `string` type does not have the same methods for concatenation, so basically such a constraint would not allow you to use `+` between the operands because the compiler has no idea how to translate that to IL and the runtime cannot know that any specific method would exist on the type `T`.

**Reply by ghost** [reply]
> Your suggestion regarding addition and concatenation would not work. The constraint `IAdditionOperators<T,T,T>` informs both the compiler and the runtime about what specific methods exist on the type that enables the use of addition. The `string` type does not have the same methods for concatenation, so basically such a constraint would not allow you to use `+` between the operands because the compiler has no idea how to translate that to IL and the runtime cannot know that any specific method would exist on the type `T`.

I take what you said.

But it just like what we can see in something like "GetEnumerator()" or "GetAsyncEnumerator()"——You don't need to implement "IEnumerable" or"IAsyncEnumerable", but just with the functions in the class, the IL will know them smartly. So it's also for the example given by me above (Maybe to check whether you've got the implementaion of "+" ?) without using the interface directly....

**Reply by ghost** [reply]
Maybe we can take the smart intellisense with IL to avoid boxing/unboxing....ect to make a wider range.

**Reply by ghost** [reply]
In fact, if I say something like：

```csharp
public static T GenericAdd<T>(T t1,T t2) where T : (IAdditionOperators<T,T,T> or string)
        {
            //Here T will be either implementation of IAdditionOperators<T,T,T> or string, but if "string" is another class that doens't implement "+", the compiler will tell you that it doesn't work by VS's intellisense
            return t1 + t2;  
       }
```

**Reply by svick** [reply]
> the IL will know them smartly.

That's not how it works. When you use a `foreach`, it's the C# compiler that finds the `GetEnumerator` method, not IL. The mechanisms in IL to resolve which method to call are:

* `virtual`/`interface` dispatch
* `dynamic`

And neither of them is suitable here.

**Reply by ghost** [reply]
> > the IL will know them smartly.
> 
> That's not how it works. When you use a `foreach`, it's the C# compiler that finds the `GetEnumerator` method, not IL. The mechanisms in IL to resolve which method to call are:
> 
> * `virtual`/`interface` dispatch
> * `dynamic`
> 
> And neither of them is suitable here.

So why not make C# compiler be sensible to the operator? Just like the functions....
You see "dynamic" can be sensible for the C# compiler, it should be easy for the compiler to know whether a generic type has the "+" (any operators or other things).....

I mean if you define something like this, this should be very similar to a template. And inside you can use "Pattern" to decide which type it is.

```csharp
public static T GenericAdd<T>(T t1,T t2) where T : (IAdditionOperators<T,T,T> or string)
        {
            //Here T will be either implementation of IAdditionOperators<T,T,T> or string, but if "string" is another class that doens't implement "+", the compiler will tell you that it doesn't work by VS's intellisense
            return t1 + t2;  
       }
```
What's more：now the key point is：if you write something like 
```csharp
GenericAdd<int> or GenericAdd<string>
```
the compiler will check whether the specific kind of type has implemented the operator or not.

But if for something like:
```csharp
GenericAdd<Array>  //Wrong, intellisense will tell you that it won't work, because "Arrary" has implemented the operator "+" yet.
```

The advantages are：

1）Avoid boxing/unboxing from object to a certain type, if both of them support "+” in the practical example.
2）Pattern for the type limitations to enhance nicer functions.

PS：This is my junior idea about the Language suggestions, maybe the final won't be something like it, but I still hope there's something more to discuss or change about it :)

**Reply by CyrusNajmabadi** [reply]
I still don't know what a real world example is where this would be useful.

**Reply by HaloFour** [reply]
The example of `+` isn't a great one because that doesn't exist in the runtime.  It exists in the C# compiler, but it can mean one of several different things.  In the case of `int` it means the IL `add` opcode.  In the case of `string` it means the `String.Concat` method, or perhaps constant folding.  In the case of custom types or the `IAdditionOperators<T1, T2, TR>` interface it means a static method that has the naming convention `op_Addition`.  The C# compiler has to choose one of those implementations and emit that as the IL in the body of this generic method.
